sycl/include/syclcompat/launch_experimental.hpp

```diff
@@ -0,0 +1,390 @@
+ *  Exceptions. See https://llvm.org/LICENSE.txt for license information.
+ *  SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  SYCLcompat
+ *
+ *  launch_experimental.hpp
+ *
+ *  Description:
+ *    Provides the interface for launching a kernel with launch properties as
+ *    well as kernel properties, using the enqueue_functions extensions for
+ *    launching the kernel.
+ *    Depends on the following 3 extensions-
+ *       sycl_ext_oneapi_kernel_properties
+ *       sycl_ext_oneapi_enqueue_functions
+ *       sycl_ext_oneapi_properties
+ **************************************************************************/
+
+#pragma once
+
+#include <sycl/ext/oneapi/experimental/enqueue_functions.hpp>
+#include <sycl/ext/oneapi/properties/properties.hpp>
+
+#include <sycl/sycl.hpp>
+
+#include <tuple>
+
+#include <syclcompat/device.hpp>
+#include <syclcompat/dims.hpp>
+#include <syclcompat/kernel_properties.hpp>
+#include <syclcompat/launch.hpp>
+
+#if defined(SYCL_EXT_ONEAPI_KERNEL_PROPERTIES) &&                              \
+    defined(SYCL_EXT_ONEAPI_PROPERTIES)
+// defined(SYCL_EXT_ONEAPI_ENQUEUE_FUNCTIONS) uncomment once
+// SYCL_EXT_ONEAPI_ENQUEUE_FUNCTIONS is defined
+
+namespace sycl_exp = sycl::ext::oneapi::experimental;
+
+namespace syclcompat {
+namespace experimental {
+namespace detail {
+
+template <auto KernelFunc, typename tuple, std::size_t... I>
+__attribute__((always_inline)) inline void
+run_kernel(tuple args, std::index_sequence<I...>) {
+  KernelFunc(std::get<I>(args)...);
+}
+
+template <auto KernelFunc, typename tuple>
+__attribute__((always_inline)) inline void run_kernel(tuple args) {
+  auto indices = std::make_index_sequence<std::tuple_size_v<tuple>>{};
+  run_kernel<KernelFunc>(args, indices);
+}
+
+template <auto KernelFunc, class KernelPropertiesStruct, bool UsesLocalMemory,
+          typename... Args>
+struct KernelFunctor {
+  KernelFunctor(Args... args, char *local_mem_ptr = nullptr)
+      : argument_tuple(std::make_tuple(args...)), local_mem_ptr(local_mem_ptr) {
+  }
+
+  auto get(sycl_exp::properties_tag) { return kernel_properties; }
+
+  __attribute__((always_inline)) inline void
+  operator()(sycl::nd_item<3> it) const {
+    if constexpr (UsesLocalMemory) {
+      run_kernel<KernelFunc>(
+          std::tuple_cat(argument_tuple, std::make_tuple(local_mem_ptr)));
+    } else {
+      run_kernel<KernelFunc>(argument_tuple);
+    }
+  }
+
+  std::tuple<Args...> argument_tuple;
+  char *local_mem_ptr;
+  static constexpr auto kernel_properties =
+      KernelPropertiesStruct::kernel_properties;
+};
+} // namespace detail
```

Comment:
Instead of trying to wrap the kernel function and all the kernel attributes in this internal KernelFunctor struct, wouldn't it be simpler (and more flexible) to allow the caller to pass a KernelFunctor "like" struct as a parameter for the launch function? Something in the line of:

```
template <auto KernelFunctor, typename... Args>
launch(const sycl::nd_range<3> &launch_params, KernelFunctor kernelFunctor,
       const sycl::queue &queue, Args... args) { 
       ...
}
```

This should allow you to simplify this code a lot

Reply:
The `launch`  which are not in the detail namespace, are the user facing `launch`'s, which will  be called by the user. Since the KernelFunctor struct is a requirement of the extension, I do not suppose it should be passed on to the user. Also I wanted to keep it similar to the current launch APIs,

Also, how would it make this more flexible, I did not get that part, so if you could please elaborate

Reply:
Users can implement the struct in whatever way they want and provide whatever list of kernel properties they need. They just have to maintain the signature of KernelFunctor so the launcher knows which methods to call. That's super flexible from the user's point of view.

In this PR, you're essentially asking the user for each individual piece of information in the KernelFunctor struct so you can build your own internal KernelFunctor. That's forcing you to define over 20 new launch functions to cover all the combinations.

If the user provides the KernelFunctor, you can mostly reuse the current launch API. Just add a new parameter (the kernelFunctor) and replace F (the function kernel) template parameter with KernelFunctor. The rest of the API remains the same.

Reply:
I think this is a fair compromise: if the user wants to do more complicated stuff, they can be responsible for creating the KernelFunctor. Can you do this in the `syclcompat::experimental` namespace still, till we figure out the best long term stable solution?